crates/egui_extras/src/loaders/gif_loader.rs

```diff
@@ -0,0 +1,135 @@
+use egui::{
+    ahash::HashMap,
+    decode_gif_uri,
+    load::{Bytes, BytesPoll, ImageLoadResult, ImageLoader, ImagePoll, LoadError, SizeHint},
+    mutex::Mutex,
+    ColorImage, GifFrameDurations, Id,
+};
+use image::AnimationDecoder as _;
+use std::{io::Cursor, mem::size_of, sync::Arc, time::Duration};
+
+/// Array of Frames and the duration for how long each frame should be shown
+#[derive(Debug, Clone)]
+pub struct AnimatedImage {
+    frames: Vec<Arc<ColorImage>>,
+    frame_durations: GifFrameDurations,
+}
+
+impl AnimatedImage {
+    pub fn byte_len(&self) -> usize {
+        size_of::<Self>()
+            + self
+                .frames
+                .iter()
+                .map(|image| {
+                    image.pixels.len() * size_of::<egui::Color32>() + size_of::<Duration>()
+                })
+                .sum::<usize>()
+    }
+
+    /// Gets image at index
+    pub fn get_image(&self, index: usize) -> Arc<ColorImage> {
+        self.frames[index % self.frames.len()].clone()
+    }
+}
+type Entry = Result<Arc<AnimatedImage>, String>;
+
+#[derive(Default)]
+pub struct GifLoader {
+    cache: Mutex<HashMap<String, Entry>>,
+}
+
+impl GifLoader {
+    pub const ID: &'static str = egui::generate_loader_id!(GifLoader);
+}
+
+pub fn gif_to_sources(data: Bytes) -> Result<AnimatedImage, String> {
+    let decoder = image::codecs::gif::GifDecoder::new(Cursor::new(data))
+        .map_err(|_err| "Couldnt decode gif".to_owned())?;
+    let mut images = vec![];
+    let mut durations = vec![];
+    for frame in decoder.into_frames() {
+        let frame = frame.map_err(|_err| "Couldnt decode gif".to_owned())?;
+        let img = frame.buffer();
+        let pixels = img.as_flat_samples();
+
+        let delay: Duration = frame.delay().into();
+        images.push(Arc::new(ColorImage::from_rgba_unmultiplied(
+            [img.width() as usize, img.height() as usize],
+            pixels.as_slice(),
+        )));
+        durations.push(delay);
+    }
+    Ok(AnimatedImage {
+        frames: images,
+        frame_durations: GifFrameDurations(Arc::new(durations)),
+    })
+}
+
+impl ImageLoader for GifLoader {
+    fn id(&self) -> &str {
+        Self::ID
+    }
+
+    fn load(&self, ctx: &egui::Context, frame_uri: &str, _: SizeHint) -> ImageLoadResult {
+        let uri_index = decode_gif_uri(frame_uri).map_err(|e| LoadError::Loading(e.to_owned()));
+        let image_uri = uri_index.as_ref().map(|v| v.0).unwrap_or(frame_uri);
+        let mut cache = self.cache.lock();
+        if let Some(entry) = cache.get(image_uri).cloned() {
+            let index = uri_index?.1;
```

Comment:
It is unclear wether or not this code is intended to support only URI:s pointing to specific frames, or if it should _also_ support URI:s pointing to the whole gif.

If the latter, please add comments explaining what that means and should result in.
If the format, please simplify the code by unwrapping failure to parse the frame index right away:

Assuming this code should 

```suggestion
        let (image_uri, frame_index) = decode_gif_uri(frame_uri).map_err(|e| LoadError::Loading(e.to_owned()))?;
        let mut cache = self.cache.lock();
        if let Some(entry) = cache.get(image_uri).cloned() {
```

Reply:
LoadError::Loading can't be used because there is no check if the image is supported before. I changed it so if the pattern isn't matched it isn't a supported image